overlord/exportstate/handlers.go

```diff
@@ -0,0 +1,134 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package exportstate
+
+import (
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// doExportContent is the do handler of export-content task.
+func (m *ExportManager) doExportContent(task *state.Task, tomb *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(task)
+	defer perfTimings.Save(st)
+
+	// Compute the snap revision export state for the snap revision we are working with.
+	snapsup, err := snapstate.TaskSnapSetup(task)
+	if err != nil {
+		return err
+	}
+	info, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+	if err != nil {
+		return err
+	}
+	manifest := NewManifestForSnap(info)
+	if manifest.IsEmpty() {
+		// Most snaps do not export any content.
+		return nil
+	}
+
+	// Create exported files, removing partial state on failure.
+	if err := manifest.CreateExportedFiles(); err != nil {
+		manifest.RemoveExportedFiles()
+		return err
+	}
+
+	// Remember what we stored in the state for both undo and unexport.
+	Set(st, info.InstanceName(), info.Revision, manifest)
+	return nil
+}
+
+// undoExportContent is the undo handler of export-content task.
+func (m *ExportManager) undoExportContent(task *state.Task, tomb *tomb.Tomb) error {
+	// To undo export content we just unexport the content.
+	// XXX: storing old-manifest seems wasteful but harmless.
+	return m.doUnexportContent(task, tomb)
+}
+
+// doUnexportContent is the do handler of unexport-content task.
+func (m *ExportManager) doUnexportContent(task *state.Task, tomb *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(task)
+	defer perfTimings.Save(st)
+
+	// Get the snap revision export state for the snap revision we are working with.
+	snapsup, err := snapstate.TaskSnapSetup(task)
+	if err != nil {
+		return err
+	}
+	var manifest Manifest
+	err = Get(st, snapsup.InstanceName(), snapsup.Revision(), &manifest)
+	if err == state.ErrNoState {
+		// Most snaps do not export any content.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	if err := manifest.RemoveExportedFiles(); err != nil {
+		return err
+	}
+
+	// Forget what was exported, keeping it in task state for undo.
+	task.Set("old-manifest", &manifest)
+	Set(st, snapsup.InstanceName(), snapsup.Revision(), nil)
+	return nil
```

Comment:
because of the issue with the task own State vs what we put into the state overall, we need a task.SetStatus(DoneStatus) here.

I think this is the only one that strictly needs that, but otoh it might be more future-proof to have code like that  in all the handlers here.

Reply:
I find it surprising that a handler function can return nil and still need this. Maybe we should just fix the overlord?

Reply:
it relates to the choice that tasks by default get called without the lock taken, so we usually we get usually:

* task returns, state is written, lock is relinquished
<---  window here were we could get rebooted or lose power and the task would be run again
* taskrunner takes the lock again and sets the status

It's fixable but needs quite a few changes about what's the default flow for a task

Reply:
Thanks for explaining this. The bit of information I was missing that we indeed unlock the state between the two.

Reply:
Done, please look at a053d1efa02173a1dbcfa61280293f106f4bdae8 and see if it makes sense.